extractor.go

```diff
@@ -270,6 +270,27 @@ func extract(ctx context.Context, t Target, dst string, src archiveWalker, cfg *
 	var fileCounter int64
 	var extractionSize int64
 
+	// check if attributes should be preserved, but as non-root user
+	if _, ok := t.(*TargetDisk); ok {
+		if cfg.PreserveFileAttributes() && os.Geteuid() != 0 {
+			cfg.Logger().Warn("cannot fully preserve file attributes as non-root user: cannot set file ownership", "uid", os.Geteuid())
```

Comment:
Not for this PR, just for future discussion, I wonder if preserving all but `uid/gid` would be useful in some cases? E.g. preserving the `x` bit seems universally useful even when uid/guid can't be preserved... Unless that was already the behaviour?

Reply:
The use cases that I had in mind were in direction of forensics and handling of container images. In both cases you want to preserve as much as possible attributes from the files. 

Okay, right now, I do not preserve a set-uid-bit, but thats for now okayish imho.

However, it looks like `os.Chown` fails silently here for non-root users:
https://github.com/hashicorp/go-extract/blob/03a7cb637658d37ed89ceb56767a850b2b2d4bb3/extractor.go#L464

on my local dev system, user (`503`) and group (`20`) are similar as in te archive, but `chown` as non-root user should ™️ fail on Linux. I dont think that the github runner has the same uid/gid and is also not executed as root (`0`). I will add another test case for that.

Reply:
Ah, well :-D I wave the `chown` if we are not `uid:0` 🤦 :

https://github.com/hashicorp/go-extract/blob/f9b271092265b5352bdfa94979b4335818225c02/target_disk.go#L123-L129

Thats also the reason why I write the log entry at the beginning and not on every call of `chown`. I will prepare a make target to test via `sudo` so that we can explicitly test and assert on that.

Reply:
Oky, I invented a new `Makefile` target: `make sudo_test` and integrated it in the testing workflow. The changes can be found in commit https://github.com/hashicorp/go-extract/pull/106/commits/7fa9d28040ab3fd02221ccf5954b96b8164a726a

Reply:
Furthermore, I relocated the unix dedicated testing code in a separate test file via https://github.com/hashicorp/go-extract/pull/106/commits/8498e88bfc0900766d4bd15cee6b3400093e7d29

Reply:
> OK, I approved too early as I hadn't realised the implications of this fully. 
👌 

> I'm not sure that just a warning log is enough here, as that would be too easily overlooked. Before people start to rely on this behaviour, I think in the short term it may be better to be conservative and fail the extraction for non-root users.

Valid remark, I understand your point and it make sense to me 👌 Another angle on that is that the implementation on non-Unix systems should also fail, bc/ the unix concept of ownership does not apply to e.g. Windows systems. 

I definitely favour the idea of requiring `root/Uid:0` in the first place and fail if thats not the case, or directly fail if the platform does not support adjusting ownership instead of silently write log messages. 

The only point is that if users adopt running the library as root to preserve ownership and then we relocate the flag for ownership into another option, users might loose that option. 

Anyhow, thats something for the release notes. 

Furthermore, the more I think about inventing further extraction options, I am afraid making the library another :can-of-worms: 🙈 However, incase of an extraction target thats not `TargetDisk`, e.g., `TargetMemory`, running the library and enable [`preserveFileAttributes`](https://github.com/hashicorp/go-extract/blob/9ab3635987c7fd3a334e0f404e66c91873a22046/config.go#L81) does not require `root/uid:0` to set ownership. See:

https://github.com/hashicorp/go-extract/blob/9ab3635987c7fd3a334e0f404e66c91873a22046/target_memory.go#L276-L291

Reply:
Mhm ... looking into the implementation of go-slug, the ownership option is ignored while restoring file attributes:
https://github.com/hashicorp/go-slug/blob/53c172a3e00fc1ac280365255b56af57846abdb4/internal/unpackinfo/unpackinfo.go#L147-L156

Reply:
I am referencing go-slug, bc/ thats the functionality we want to replace with go-extract in the near future.

Reply:
^ that given and a bunch of thinking brought me to the point that it make sense to have two dedicated flags to preserve ownership and file attributes. 

I added that in https://github.com/hashicorp/go-extract/pull/106/commits/00f38ebd6e55448bd8bf961f7add882ef1d2d4ae . 

Sorry, that I stepped over you draft PR #111 🙈

Reply:
Discussion continued in https://github.com/hashicorp/go-extract/pull/106#discussion_r1887362198